tlog-witness.md

```diff
@@ -0,0 +1,226 @@
+[RFC 8174]: https://www.rfc-editor.org/rfc/rfc8174.html
+[RFC 6962]: https://www.rfc-editor.org/rfc/rfc6962.html
+
+## Introduction
+
+This protocol allows clients, usually logs, to obtain cosignatures from
+witnesses, making it possible to produce self-contained inclusion proofs that
+can be verified offline. When producing a new checkpoint, the log reaches out to
+witnesses to request cosignatures over it, providing a consistency proof.
+Witnesses verify that the checkpoint is consistent with the previous state of
+the tree, and return a timestamped cosignature.
+
+A witness is an entity exposing an HTTPS service identified by a name and a
+public key. Each witness is configured with a list of supported log public keys.
+For each log, uniquely identified by its origin line, the witness keeps only
+track of the latest checkpoint it observed and verified.
+
+Clients are not expected to communicate directly with the witnesses, logs and
+(sometimes) monitors are, but there is no authentication of requests beyond the
+validation of the signature on the checkpoint.
+
+Witnesses need to be operated by trusted, reliable entities with minimal churn.
+Log deployments pick a witness trust policy that gets deployed into clients and
+might be difficult or impossible to change. Developing a healthy ecosystem of
+witnesses might be the biggest hurdle to widespread tlog success. It’s
+impractical to build a separate witness ecosystem for each application, so it’s
+important that witnesses be public, interoperable, and scalable. As for
+quantity, there should be more than a few, but once reached a critical mass each
+additional witness has diminishing marginal value. Additional log deployments,
+on the other hand, don’t have diminishing returns and are a desirable goal.
+Because of this, witnesses are designed to scale well with a large number of
+rarely active logs, and to support diverse log designs, including low-latency
+and "serverless" logs.
+
+## HTTPS Interface
+
+A witness is defined by a name, a public key, and by two URL prefixes: the
+*submission prefix* for write APIs and the *monitoring prefix* for read APIs.
+
+A witness MAY use the same value for both the *submission prefix* and the
+*monitoring prefix*.
+
+If exposing the machine holding the witness key material directly to the
+Internet is undesirable, operators MAY use a [bastion][].
+
+### add-checkpoint
+
+The `add-checkpoint` call is used to submit a new checkpoint to the witness,
+along with a consistency proof from a previous checkpoint, and returns the
+cosignature.
+
+    POST <submission prefix>/add-checkpoint
+
+The request MUST be an HTTP POST. Clients SHOULD use and witnesses SHOULD
+support keep-alive HTTP connections to reduce latency and load due to connection
+establishment.
+
+The request body MUST be a sequence of
+  - a previous size line,
+  - zero or more consistency proof lines,
+  - and an empty line,
+  - followed by a [checkpoint][].
+
```

Comment:
Will this consistency proof format be extracted into the proposed `tlog-offline-proof` spec? Would be great to have consistent (haha) consistency and inclusion proof formats.